src/satosa/backends/saml2.py

```diff
@@ -35,6 +35,31 @@
 logger = logging.getLogger(__name__)
 
 
+def get_memorized_idp(context, config, force_authn):
+    memorized_idp = (
+        config.get(SAMLBackend.KEY_MEMORIZE_IDP)
+        and context.state.get(Context.KEY_MEMORIZED_IDP)
+    )
+    use_when_force_authn = config.get(
+        SAMLBackend.KEY_USE_MEMORIZED_IDP_WHEN_FORCE_AUTHN
+    )
+    value = (not force_authn or use_when_force_authn) and memorized_idp
+    return value
+
+
+# XXX check KEY_FORCE_AUTHN value type (boolean vs str)
+def get_force_authn(context, config, sp_config):
+    value = (
+        config.get(SAMLBackend.KEY_MIRROR_FORCE_AUTHN)
+        and (
+            context.state.get(Context.KEY_FORCE_AUTHN)
+            or context.get_decoration(Context.KEY_FORCE_AUTHN)
+        )
+        or sp_config.getattr("force_authn", "sp")
```

Comment:
`force_authn` might be set to `True` through the _pysaml2 configuration_ for this SP. We need to take that into account when deciding if we're going to redirect to the discovery service or use the memorized-idp (when set)